### DISCUSSION TITLE
No negative control pair passes pairwise QC.

### DISCUSSION THREAD

**Discussion by AAAAAbetter** [Katsevich-Lab/sceptre#155] [Q&A] [answered]
Your package has been extremely helpful!

However, when running the code, I encountered the following error:  
"Constructing negative control pairs.  
Error in construct_negative_control_pairs_v2(sceptre_object = sceptre_object, :  
No negative control pair passes pairwise QC."

My original code is as follows:
```
sceptre_object <- import_data_from_cellranger(
  directories = directories,
  moi = "high",
  grna_target_data_frame=grna_target_data_frame_highmoi
)

positive_control_pairs <- construct_positive_control_pairs(sceptre_object)

discovery_pairs<- construct_trans_pairs(
  sceptre_object = sceptre_object,
  positive_control_pairs = positive_control_pairs,
  pairs_to_exclude = "pairs_containing_pc_targets"
)

sceptre_object <- sceptre_object |>
  set_analysis_parameters(
    discovery_pairs = discovery_pairs,
    positive_control_pairs = positive_control_pairs,
    side = "left",
    grna_integration_strategy = "singleton",
    multiple_testing_alpha = 0.2
  ) |>
  assign_grnas(parallel = T,n_processors=2) |>
  run_qc(p_mito_threshold = 0.075) |>
  run_calibration_check(parallel = T, n_processors=2) |>
  run_power_check(parallel = T,n_processors=2) |>
  run_discovery_analysis(parallel = T, n_processors=2)
```

After adjusting `grna_integration_strategy = "singleton"` to the default option, the code runs without issues. Could you please explain the potential reasons for this error? What adjustments can be made during the experiment and sequencing to avoid this error?

**Comment by timothy-barry** [comment]
Good to hear!

The issue that you are encountering relates to pairwise QC. A negative control pair is analyzed only if it passes [pairwise QC](https://timothy-barry.github.io/sceptre-book/run-qc.html#sec-qc_pairwise_qc). You can tinker with the pairwise QC thresholds by setting `n_nonzero_trt` and `n_nonzero_cntrl` in `set_analysis_parameters()`. The default value of these parameters is `7`; decreasing this value will make pairwise QC more liberal, thus increasing the chance that a negative control pair will pass pairwise QC.

If you don't mind, would you mind telling me a bit more about your experiment? In particular, how many negative control gRNAs do you have? How many targeting gRNAs do you have? Any how many total targets do you have? (In other words, how many targeting gRNAs per target site)?

**Reply by AAAAAbetter** [reply]
Thanks for your reply!
I have a total of 6000 gRNAs in my library, there are about 1200 genes in this library, an average of 5 sgRNAs per gene, and 31 gRNAs are non-targeted gRNAs.

**Comment by ekatsevi** [comment, marked as answer]
Thanks for the additional details, and apologies for not following up sooner. Your original run used `grna_integration_strategy = "singleton"`, whereas the default is `"union"`. Under singleton, each negative-control group contains one non-targeting sgRNA; under union, approximately five sgRNAs are pooled per group in your experiment. The pooled groups contain more treatment cells, explaining why union passes pairwise QC while singleton does not.

Thus, the error means that no individual non-targeting sgRNA–response pair met the default pairwise-QC thresholds; it does not necessarily mean that 31 non-targeting sgRNAs is too few. If singleton analysis is required, the main remedies are more usable cells or expression coverage per sgRNA, or cautiously lowering `n_nonzero_trt_thresh` and `n_nonzero_cntrl_thresh` in `run_qc()`.

Since switching back to union resolved the problem, I’m marking this as answered.
